tools/data-handler/src/macros/BaseMacro.ts

```diff
@@ -0,0 +1,118 @@
+/**
+    Cyberismo
+    Copyright © Cyberismo Ltd and contributors 2024
+
+    This program is free software: you can redistribute it and/or modify it under the terms of the GNU Affero General Public License version 3 as published by the Free Software Foundation.
+
+    This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Affero General Public License for more details.
+
+    You should have received a copy of the GNU Affero General Public
+    License along with this program.  If not, see <https://www.gnu.org/licenses/>.
+*/
+
+import { MacroGenerationContext, MacroMetadata } from './common.js';
+import { handleMacroError } from './index.js';
+
+abstract class BaseMacro {
+  /**
+   * Used for identifying the macro globally.
+   * This should be async safe.
+   */
+  private static globalCounter = 0;
+
+  private globalId: string;
+  private promiseList: {
+    localId: number;
+    promise: Promise<void>;
+    placeholder: string;
+    promiseResult: string | null;
+  }[] = [];
+
+  constructor(protected macroMetadata: MacroMetadata) {
+    // Set a unique global id for each instance of the macro, based on the macro name
+    this.globalId = `${this.macroMetadata.name}-${BaseMacro.globalCounter++}`;
+  }
+
+  protected abstract handleStatic(
+    context: MacroGenerationContext,
+    input: string,
+  ): Promise<string>;
+
+  protected abstract handleInject(
+    context: MacroGenerationContext,
+    input: string,
+  ): Promise<string>;
+
+  public get metadata() {
+    return this.macroMetadata;
+  }
+
+  /**
+   * Function responsible for starting the promise and storing it along with its localId.
+   */
+  public handleMacro = (context: MacroGenerationContext, input: string) => {
```

Comment:
Not a big fan of functions that have `handle` (or `manage`) in their name. It is usually programmer mumbojambo that doesn't mean a thing.
Could this be called `storeMacro` or `prepareMacro`? 
And then when calling the macro - instead of `handleResult`, it could be called `execute`.